File: python/src/nnabla/variable.py
# Copyright (c) 2017 Sony Corporation. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import absolute_import

from ._variable import Context
from ._variable import Variable as _Variable


# As Cython does not support some arithmetic operations,
# Variable is inherited in order to define arithmetic operations.
class Variable(_Variable):
    __doc__ = _Variable.__doc__

    def __repr__(self):
        return "<Variable({}, need_grad={}) at {}>".format(
            self.shape, self.need_grad, hex(id(self)))

    def __add__(self, other):
        """
        Element-wise addition.
        Implements the addition operator expression ``A + B``, together with :func:`~nnabla.variable.__radd__` .
        When a scalar is specified for ``other``, this function performs an
        element-wise operation for all elements in ``self``.

        Args:
            other (float or ~nnabla.Variable): Internally calling
                :func:`~nnabla.functions.add2` or
                :func:`~nnabla.functions.add_scalar` according to the
                type.

        Returns: :class:`nnabla.Variable`

        """
        import nnabla.functions as F
        if isinstance(other, Variable):
            return F.add2(self, other)
        return F.add_scalar(self, other)

    def __mul__(self, other):
        """
        Element-wise multiplication.
        Implements the multiplication operator expression ``A * B``, together with :func:`~nnabla.variable.__rmul__` .
        When a scalar is specified for ``other``, this function performs an
        element-wise operation for all elements in ``self``.

        Args:
            other (float or ~nnabla.Variable): Internally calling
                :func:`~nnabla.functions.mul2` or
                :func:`~nnabla.functions.mul_scalar` according to the
                type.

        Returns: :class:`nnabla.Variable`

        """
        import nnabla.functions as F
        if isinstance(other, Variable):
            return F.mul2(self, other)
        return F.mul_scalar(self, other)

    def __sub__(self, other):
        """
        Element-wise subtraction.
        Implements the subtraction operator expression ``A - B``, together with :func:`~nnabla.variable.__rsub__` .
        When a scalar is specified for ``other``, this function performs an
        element-wise operation for all elements in ``self``.

        Args:
            other (float or ~nnabla.Variable): Internally calling
                :func:`~nnabla.functions.sub2` or
                :func:`~nnabla.functions.add_scalar` according to the
                type.

        Returns: :class:`nnabla.Variable`

        """
        import nnabla.functions as F
        if isinstance(other, Variable):
            return F.sub2(self, other)
        return F.add_scalar(self, -other)

    def __rsub__(self, other):
        """
        Element-wise subtraction.
        Part of the implementation of the subtraction operator.

        Args:
            other (float or ~nnabla.Variable): Internally calling
                :func:`~nnabla.functions.sub2` or
                :func:`~nnabla.functions.r_sub_scalar` according to the
                type.

        Returns: :class:`nnabla.Variable`

        """
        import nnabla.functions as F
        if isinstance(other, Variable):
            return F.sub2(other, self)
        return F.r_sub_scalar(self, other)

    def __div__(self, other):
        """
        Element-wise division.
        Implements the division operator expression ``A / B``, together with :func:`~nnabla.variable.__rdiv__` .
        When a scalar is specified for ``other``, this function performs an
        element-wise operation for all elements in ``self``.

        Args:
            other (float or ~nnabla.Variable): Internally calling
                :func:`~nnabla.functions.div2` or
                :func:`~nnabla.functions.mul_scalar` according to the
                type.

        Returns: :class:`nnabla.Variable`

        """
        import nnabla.functions as F
        if isinstance(other, Variable):
            return F.div2(self, other)
        return F.mul_scalar(self, 1. / other)

    def __rdiv__(self, other):
        """
        Element-wise division.
        Part of the implementation of the division operator.

        Args:
            other (float or ~nnabla.Variable): Internally calling
                :func:`~nnabla.functions.sub2` or
                :func:`~nnabla.functions.add_scalar` according to the
                type.

        Returns: :class:`nnabla.Variable`

        """
        import nnabla.functions as F
        if isinstance(other, Variable):
            return F.div2(other, self)
        return F.r_div_scalar(self, other)

    def __pow__(self, other):
        """
        Element-wise power function.
        Implements the power operator expression ``A ** B``, together with :func:`~nnabla.variable.__rpow__` .
        When a scalar is specified for ``other``, this function performs an
        element-wise operation for all elements in ``self``.

        Args:
            other (float or ~nnabla.Variable): Internally calling
                :func:`~nnabla.functions.pow2` or
                :func:`~nnabla.functions.pow_scalar` according to the
                type.

        Returns: :class:`nnabla.Variable`

        """
        import nnabla.functions as F
        if isinstance(other, Variable):
            return F.pow2(self, other)
        return F.pow_scalar(self, other)

    def __rpow__(self, other):
        """
        Element-wise power function.
        Part of the implementation of the power operator expression.

        Args:
            other (float or ~nnabla.Variable): Internally calling
                :func:`~nnabla.functions.pow2` or
                :func:`~nnabla.functions.r_pow_scalar` according to the
                type.

        Returns: :class:`nnabla.Variable`

        """
        import nnabla.functions as F
        if isinstance(other, Variable):
            return F.pow2(other, self)
        return F.r_pow_scalar(self, other)

    def __radd__(self, other):
        """
        Element-wise addition.
        Part of the implementation of the addition operator expression.

        Args:
            other (float or ~nnabla.Variable): Replace with :meth:`__add__`.

        Returns: :class:`nnabla.Variable`

        """
        return self + other

    def __rmul__(self, other):
        """
        Element-wise division.
        Part of the implementation of the division operator expression.

        Args:
            other (float or ~nnabla.Variable): Replace with :meth:`__mul__`.

        Returns: :class:`nnabla.Variable`

        """
        return self * other

    def __truediv__(self, other):
        """
        Element-wise division.
        Part of the implementation of the division operator expression.
        """
        return self.__div__(other)

    def __rtruediv__(self, other):
        """
        Element-wise division.
        Part of the implementation of the division operator expression.
        """
        return self.__rdiv__(other)

    def __pos__(self):
        """
        This function simply returns itself.
        Implements the unary plus operator, ``+A``.

        Returns: :class:`nnabla.Variable`

        """
        return self

    def __neg__(self):
        """
        Element-wise negation.
        Implements the unary negation expression ``-A`` .

        Returns: :class:`nnabla.Variable`

        """
        import nnabla.functions as F
        return F.mul_scalar(self, -1)

File: src/nbla/variable.cpp
// Copyright 2017,2018,2019,2020,2021 Sony Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Variable.cpp
#include <nbla/variable.hpp>

#include <functional>
#include <memory>
#include <string>

namespace nbla {

using std::make_shared;

void BaseVariable::update_python_user_reference_counts(const int diff) {
  python_user_reference_counts += diff;
  data_->update_python_user_reference_counts(diff);
}

void BaseVariable::set_data(NdArrayPtr data) {
  NBLA_CHECK(data->shape() == shape_, error_code::value, "Shape must match.");

  // New NdArray traces the Variable tree.
  data->update_python_user_reference_counts(python_user_reference_counts);

  if (data_) {
    // Old NdArray purges the Variable tree.
    data_->update_python_user_reference_counts(-python_user_reference_counts);
  }

  data_ = data;
}

BaseVariable::~BaseVariable() {
  data_->update_python_user_reference_counts(-python_user_reference_counts);
}

void BaseVariable::set_grad(NdArrayPtr grad) {
  NBLA_CHECK(grad->shape() == shape_, error_code::value, "Shape must match.");
  grad_ = grad;
}

void Variable::update_shape_info() {
  size_ = compute_size_by_shape(shape_);
  strides_ = get_c_contiguous_strides(shape_);
  ndim_ = shape_.size();
}

Variable::Variable(const Shape_t &shape) {
  this->shape_ = shape;
  update_shape_info();
  this->set_data(make_shared<NdArray>(shape_));
  this->set_grad(make_shared<NdArray>(shape_));
}

Variable::Variable(NdArrayPtr data) {
  shape_ = data->shape();
  update_shape_info();
  this->set_data(data);
  this->set_grad(make_shared<NdArray>(shape_));
}

void Variable::reshape(const vector<int64_t> &shape, bool force) {
  if (shape_ == shape)
    return;
  const Size_t size = compute_size_by_shape(shape);
  if (size_ == size) {
    shape_ = shape;
    update_shape_info();
    this->data()->reshape(shape);
    this->grad()->reshape(shape);
    return;
  }
  NBLA_CHECK(force, error_code::value,
             "Total dimensions not match. Set force=true if you want to "
             "resize array (clearing data). Given: %d != current: %d.",
             size, size_);
  shape_ = shape;
  update_shape_info();
  this->data()->reshape(shape_, true);
  this->grad()->reshape(shape_, true);
}

VariablePtr Variable::view() {
  auto v = make_shared<Variable>(shape_);
  v->set_data(this->data());
  v->set_grad(this->grad());
  return v;
}

VariablePtr Variable::view(const Shape_t &shape) {
  const Size_t size = compute_size_by_shape(shape);
  NBLA_CHECK(size == size_, error_code::value,
             "The total size must be the same as the variable. "
             "Given: %d != current: %d.",
             size, size_);
  auto v = make_shared<Variable>(shape);
  v->set_data(this->data()->view(shape));
  v->set_grad(this->grad()->view(shape));
  return v;
}

Size_t Variable::size(Size_t axis) const {
  if (axis <= 0) {
    return size_;
  }
  return compute_size_by_shape(shape_, axis);
}
}
